### DISCUSSION TITLE
Use dependency in functions

### DISCUSSION THREAD

**Discussion by klaydev** [fastapi/fastapi#10503] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends

from mymodule import get_some_dependency, SomeDependecyClass

async def not_route(sd: SomeDependencyClass = Depends(get_some_dependency): # ???
    ...
```

### Description

Hello everyone. The question may seem silly, but is it possible to use dependencies in simple functions? It would be very useful to use them not only in routes, but also in ordinary functions

### Operating System

Linux, Windows

### FastAPI Version

0.104

### Pydantic Version

2

### Python Version

3.11.6

**Comment by White-Mask** [comment]
Hello @KlaydKladowayIgr , looking at your example code and questions, I will now respond step by step based on my experience.

1. Is it possible to use dependencies in simple functions? Yes, it is possible to use dependencies in FastAPI. Below, I will show you a simple example of an F1 API:
([Dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/))
```python
from fastapi import FastAPI, Depends, HTTPException

app = FastAPI(
    title="F1 Drivers API",
    version="1.0.0",
    description="API to manage F1 drivers",
    root_path="/api/v1",
)


class MockDatabase:
    def __init__(self):
        self.race_data = {
            1: {"name": "Italian Grand Prix", "year": 2023},
            2: {"name": "Spanish Grand Prix", "year": 2023},
        }
        self.driver_data = {
            1: {"name": "Lewis Hamilton", "team": "Mercedes"},
            2: {"name": "Max Verstappen", "team": "Red Bull Racing"},
        }

    def get_race(self, race_id: int):
        return self.race_data.get(race_id)

    def get_driver(self, driver_id: int):
        return self.driver_data.get(driver_id)


def mock_database():
    return MockDatabase()


@app.get("/race/{race_id}")
async def get_race_details(race_id: int, db: MockDatabase = Depends(mock_database)):
    race = db.get_race(race_id)
    if not race:
        raise HTTPException(status_code=404, detail="Race not found")
    return race


@app.get("/drivers/{driver_id}")
async def get_driver(driver_id: int, database: MockDatabase = Depends(mock_database)):
    driver = database.get_driver(driver_id)
    if not driver:
        raise HTTPException(status_code=404, detail="Driver not found")
    return driver
```

As you can see, we use dependencies in routes to inject common functionalities, such as database connection in this case. This helps us avoid code repetition. 

2. Dependencies can also be used in functions, as demonstrated in the following example: 
* ([Sub-dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/))
```python
class MockDatabase:
    ...
    def get_race(self):
        return self.race_data

def validate_year(year: int):
    if year < 2023:
        raise HTTPException(status_code=400, detail="Year must be 2023")

def get_races_by_year(year: int, db: MockDatabase = Depends(mock_database)):
    race_data = db.get_race()
    filtered_races = [race for race in race_data.values() if race['year'] == year]
    if not filtered_races:
        raise HTTPException(status_code=404, detail="Race not found")
    return filtered_races

@app.get("/races/{year}", dependencies=[Depends(validate_year)])
async def get_races(races: list = Depends(get_races_by_year)):
    return races
````

3. Dependencies can also be used like decorators, example:
* ([Dependencies in path operation decorators](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/))
```python
class MockDatabase:
    ...
    def get_races(self, year: int):
        return [race for race in self.race_data.values() if race['year'] == year]

def validate_year(year: int):
    if year != 2023:
        raise HTTPException(status_code=400, detail="Year must be 2023")

@app.get("/races/{year}", dependencies=[Depends(validate_year)])
async def get_race_by_year(year: int, db: MockDatabase = Depends(mock_database)):
    races = db.get_races(year)
    if not races:
        raise HTTPException(status_code=404, detail="Race not found")
    return races
```

This approach offers several significant advantages:

* Database Connections: You can inject dependencies to manage database connections, simplifying the handling of critical resources and ensuring efficient usage.

* Authentication/Security Requirements: Authentication and security validation can be managed in an organized and centralized manner through dependencies. This not only strengthens the security of your application but also enhances code readability.

* Clients for Interacting with Other Services: Dependencies allow you to create clients for interacting with external services, streamlining communication with external APIs and third-party services.

* Code Sharing: Common or shared logic can be encapsulated within dependencies, facilitating its reuse in different parts of your application and maintaining clean and organized code.

If you're looking for alternatives or similar tools to manage dependency injection in Python, you can consider packages like [Dependency Injector](https://python-dependency-injector.ets-labs.org/) and [Injector](https://pypi.org/project/injector/). These libraries offer flexible and effective options for handling dependencies in your code.

In summary, dependency injection with FastAPI enhances the maintainability, security, and efficiency of your code by reducing repetition and improving readability. Moreover, its ability to handle nested dependencies makes it a powerful tool for building robust and scalable applications

**Comment by nymous** [comment, marked as answer]
Hello!

No you cannot use FastAPI dependencies in normal functions. The `Depends` from FastAPI are only resolved (recursively) when used in an API route.
If you want to use dependency injection in other functions you will need to code it yourself or use a separate, "generic" dependency injection library. This list could be a good resource https://github.com/sfermigier/awesome-dependency-injection-in-python

See also #1873, #2967, #4893

**Reply by White-Mask** [reply]
@nymous, my apologies for the confusion. I reviewed my response and noticed that I made some errors in interpreting the question. I have already corrected them. Regards.

**Comment by Tubaher** [comment]
Well, `Depends` is resolved by FastAPI when all the path functions (routes) are initialized. Then, if you use `Depends` in normal functions, FastAPI is not aware of them, and they are not resolved. However, at the end of the day, `Depends` is just a tool provided to manage DI. Then, you can try some similar tools (packages) like https://python-dependency-injector.ets-labs.org/ . Or for things like inject DB using `yield`, you can use the context manager decorator as function in a `with` statement https://stackoverflow.com/questions/66558958/fastapi-dependencies-yield-how-to-call-them-manually .
